zict/async_buffer.py

```diff
@@ -0,0 +1,157 @@
+        called on one of them before the actual data is fetched. ``__setitem__`` also
+        internally calls ``__delitem__`` in a non-atomic way, so you may get
+        ``KeyError`` when updating a value too.
+        """
+        # This block avoids spawning a thread if keys are missing from both fast and
+        # slow. It is otherwise just a performance optimization.
+        if missing == "omit":
+            keys = [key for key in keys if key in self]
+        elif missing == "raise":
+            for key in keys:
+                if key not in self:
+                    raise KeyError(key)
+        else:
+            raise ValueError(f"missing: expected raise or omit; got {missing}")
+        # End performance optimization
+
+        try:
+            # Do not pull keys towards the top of the LRU unless they are all available.
+            # This matters when there is a very long queue of async_get futures.
+            d = self.fast.get_all_or_nothing(keys)
+        except KeyError:
+            pass
+        else:
+            f: asyncio.Future[dict[KT, VT]] = asyncio.Future()
+            f.set_result(d)
+            return f
+
+        def _async_get() -> dict[KT, VT]:
+            d = {}
+            for k in keys:
+                if self.fast.closed:
+                    raise asyncio.CancelledError()
+                try:
+                    # This can cause keys to be restored and older keys to be evicted
+                    d[k] = self[k]
+                except KeyError:
+                    # Race condition: key was there when async_get was called, but got
+                    # deleted afterwards.
+                    if missing == "raise":
+                        raise
+            return d
+
+        return self._offload(_async_get)
+
+    def __setitem__(self, key: KT, value: VT) -> None:
+        """Immediately set a key in fast. If this causes the total weight to exceed n,
+        asynchronously start moving keys from fast to slow in a worker thread.
+        """
+        self.set_noevict(key, value)
+        self.async_evict_until_below_target()
+
+    @locked
+    def async_evict_until_below_target(self, n: float | None = None) -> None:
+        """If the total weight exceeds n, asynchronously start moving keys from fast to
+        slow in a worker thread.
+        """
+        if n is None:
+            n = self.n
+        n = max(0.0, n)
+        weight = self.fast.total_weight
+        if self.evicting is not None:
+            weight = min(weight, self.evicting)
+        if weight <= n:
+            return
+
+        self.evicting = n
+
+        def done(_: asyncio.Future) -> None:
+            self.evicting = None
+
+        # Note: this can get cancelled by LRU.close(), which in turn is
+        # triggered by Buffer.close()
+        future = self._offload(self.evict_until_below_target, n)
+        future.add_done_callback(done)
```

Comment:
IIUC this is possibly called multiple times (even if it is locked) but the done callbacks of concurrently running `evict_untiL_below_target` would overwrite / first one would already reset evicting. Is this intentional?

Reply:
You're right, fixed